### PR TITLE
custom attachments

### PR DESCRIPTION
**Pull request by brandon-reinhart**
In progress PR adding the ability to create custom attachments that can be assigned to slots.

May optionally include the ability to create custom attachment loaders.

### CODE REVIEW

**Comment by brandon-reinhart** [comment]
I am working on the interface to modify a region attachment after it has been created. The region attachment can be empty upon creation, but to render any visible contents, the user needs to set all fields and call update_region. This should be clear and not something you can mess up.

**Comment by brandon-reinhart** [comment]
spSlot_setAttachment: It appears that setting the attachment on the slot **does not increment the attachment ref count**.

This seems strange to me, as the pointer is in use and if an attachment is set on a skin the ref count is incremented. It seems to me that an attachment that is in use by a slot should have its refcount increment and dispose called when it is no longer in use, but spine c does not do this. 

As a result, it is possible to set an attachment on a slot and then drop the attachment, causing an access violation.

**Comment by brandon-reinhart** [comment]
Another issue I am investigating is that calling Slot::set_attachment cause an immediate panic. Calling the underlying c::spSlot_setAttachment works fine.

**Comment by brandon-reinhart** [comment]
ref: https://github.com/EsotericSoftware/spine-runtimes/pull/2399
ref: https://github.com/EsotericSoftware/spine-runtimes/issues/2398

**Comment by brandon-reinhart** [comment]
An alternative to making attachment accessors mutable is to place the create attachment convenience function in RegionAttachment and have it take the loader and props as arguments.

**Comment by brandon-reinhart** [comment]
Example of the current code:
```rust
        let attachment_loader = AttachmentLoader::new_atlas_loader(&atlas);

        let region_props = rusty_spine::RegionProps {
            x: 0.0,
            y: 0.0,
            scale_x: 1.0,
            scale_y: 1.0,
            rotation: 0.0,
            width: 128.0,
            height: 128.0,
            color: Color::new_rgba(1.0, 1.0, 1.0, 1.0),
        };

        let attachment = attachment_loader
            .create_region_attachment(None, "test", "head", &region_props)
            .unwrap();

        unsafe {
            let mut slot = controller.skeleton.find_slot_mut("front-shin").unwrap();
            slot.set_attachment(Some(attachment.clone()));
        }
```

**Comment by jabuwu** [comment]
Nice work.

> spSlot_setAttachment: It appears that setting the attachment on the slot does not increment the attachment ref count.

We shouldn't be making modifications directly to `spine_c.rs`, since it's auto-generated. We will need to wait for a response to the Spine C PRs.

File: src/c_interface.rs [hunk 1]
```diff
@@ -462,6 +462,30 @@ macro_rules! c_accessor_string {
...
+                    crate::c_interface::from_c_str(std::ffi::CStr::from_ptr(self.c_ptr_ref().$c))
+                } else {
+                    ""
+                }
+            }
+        }
+
+        $(#[$($attrss)*])*
+        pub fn $rust_set(&mut self, value: String) {
+            let c_str = std::ffi::CString::new(value).expect("Null byte found in provided string");
```
Comment (jabuwu) [inline src/c_interface.rs:481]: I realize it's an exceedingly rare case, but I think we should bubble up the error here, wrapped in `SpineError::NulError`.

File: src/attachment_loader.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            Self::new_from_ptr(attachment_loader)
+        }
+    }
+
+    /// Creates an [`Attachment`](`crate::Attachment`) of a specified type.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`AttachmentLoaderError::CreateAttachmentFailed`] if creating the attachment failed.
+    /// Check error1() and error2() for more information.
```
Comment (jabuwu) [inline src/attachment_loader.rs:49]:
```suggestion
    /// Check [`error1`](`Self::error1`) and [`error2`](`Self::error2`) for more information.
```

File: src/attachment_loader.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            }
+        }
+    }
+
+    /// Convenience function for creating a [`RegionAttachment`](`crate::RegionAttachment`).
+    ///
+    /// # Errors
+    ///
+    /// Returns [`AttachmentLoaderError::CreateAttachmentFailed`] if creating the attachment failed.
+    /// Check error1() and error2() for more information.
```
Comment (jabuwu) [inline src/attachment_loader.rs:91]:
```suggestion
    /// Check [`error1`](`Self::error1`) and [`error2`](`Self::error2`) for more information.
```

**Comment by brandon-reinhart** [comment]
> Nice work.
> 
> > spSlot_setAttachment: It appears that setting the attachment on the slot does not increment the attachment ref count.
> 
> We shouldn't be making modifications directly to `spine_c.rs`, since it's auto-generated. We will need to wait for a response to the Spine C PRs.

Yeah, that's why I noted this is temporary. The goal of this was to prove out it resolves the problems, which it does in my testing.

**Comment by brandon-reinhart** [comment]
This is all usable. Anything else I should do here? It doesn't require any changes from the spine runtime.

**Comment by brandon-reinhart** [comment]
Ship it!

**Pull request merged**